### DISCUSSION TITLE
Examples

### DISCUSSION THREAD

**Discussion by Aaarrrggghhhh** [canboat/canboatjs#169] [Q&A]
Hi all
Thanks heaps for the hard work here. 
I'm brand new to Node.js but have some experience with Arduino. I'm having a bit of a hard time wrapping my brain around what's required. Is there any examples anywhere which go over the various facets? Is there some documentation somewhere which describes what each module does? for example - what is the purpose of simpleCan?
I'm currently trying to read NMEA2000 & J1939 data through a pair of MCP2515 and output that information as html.  
I would be quite happy to help with the documentation if help is needed

**Comment by sbender9** [comment]
Yeah, I haven't been able to spend much time on documentation. (And there hasn't been much demand since this is mainly used for Signal K).

simpleCan allows simple programs to read from a socketcan device.

I don't know much about Arduino or MCP2515, so if you give me some more info about the data you're reading or more specific questions, I can help you along.

**Comment by Aaarrrggghhhh** [comment]
Hi Scott, thanks for commenting,

Sorry I only mentioned arduino to indicate my level of experience - not a complete moron in general but moronic in the ways of Node.
The MCP2515 is from what I read a reasonably common transciever chip used in pi hats - as mine is also. I also need to apologise, I didn't mention, I'm using a Raspberry Pi running Raspian.

My goal is to read J1939 information, send back commands through J1939 (engine start/stop, monitor NMEA2000 and send alerts over NMEA2000. Display and control via an HTML / websocket interface.

I'm at the stage where I can read NMEA2000 through socketcan, and I was hoping to use canboatjs for three functions:

- Handle the handshaking / declarations on the bus correctly 
- Decode the NMEA2000 messages to their correct information
- Provide  a structure I could follow to enter in the custom PGN's I will be using to control the J1939

Does simpleCan handle the bus addressing? How do you actually implement it? I tried:
`const SimpleCan = require('@canboat/canboatjs').SimpleCan;
const doofer = new SimpleCan()

doofer.start()

console.log('test')
`
Which Returned:
`
/home/pi/shared/Tutorial2/Test2/node_modules/@canboat/canboatjs/lib/simpleCan.js:17
  const canDevice = this.options.canDevice || 'can0'
                                 ^

TypeError: Cannot read properties of undefined (reading 'canDevice')
    at SimpleCan.start (/home/pi/shared/Tutorial2/Test2/node_modules/@canboat/canboatjs/lib/simpleCan.js:17:34)
    at Object.<anonymous> (/home/pi/shared/Tutorial2/Test2/server.js:14:8)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:79:12)
    at node:internal/main/run_main_module:17:47
`
I think there's a couple of arguments for the function SimpleCan (options, messageCb) but I don't know what these are and if I'm calling it correctly. I basically copied that code from the wiki page which uses FromPGN - but I see this has a constructor in it's js file so I'm probably barking up the wrong tree.

**Comment by sbender9** [comment]
simpleCan does handle all the handshaking, etc. And there's a way to declare custom pgns. (we can get into that once we get this working)

This is a small example of using simpleCan. I don't have a way to test this right now, so hopefully it will work!

```
const canboatjs = require('@canboat/canboatjs')
var parser = new canboatjs.FromPgn()

let messageCb = (data) => {
  let jsonData = parser.parse(data, (err) => { if ( err ) console.error(err) })
  if ( jsonData ) {
    //process
  }
}

let simpleCan = new canboatjs.SimpleCan({
  canDevice: 'can0',
  preferredAddress: 35,
  disableDefaultTransmitPGNs: true,
  transmitPGNs: [],
  addressClaim: {
    "Unique Number": 139725,
    "Manufacturer Code": 'Fusion Electronics',
    "Device Function": 130,
    "Device Class": 'Entertainment',
    "Device Instance Lower": 0,
    "Device Instance Upper": 0,
    "System Instance": 0,
    "Industry Group": 'Marine'
  },
  productInfo: {
    "NMEA 2000 Version": 1300,
    "Product Code": 667,
    "Model ID": "MS-UD650",
    "Software Version Code": "1.0",
    "Model Version": "1.0",
    "Model Serial Code": "123456",
    "Certification Level": 0,
    "Load Equivalency": 1
  }
}, messageCb)

simpleCan.start()
```

**Comment by Aaarrrggghhhh** [comment]
Wow, thanks for taking the time to leave such a detailed reply! 

I tried it out, but it didn't work til I modified simpleCan.js from line 28 - might be a bug?

looks like this:

 ```
     pgn.timestamp = new Date().toISOString()
      if ( this.plainText ) {
        messageCb(binToActisense(pgn, msg.data, msg.data.length))
      } else {
        messageCb({ pgn, length: msg.data.length, data: msg.data })

```

I changed it to: (note **this.**messageCb( ....  )

```
      pgn.timestamp = new Date().toISOString()
      if ( this.plainText ) {
        this.messageCb(binToActisense(pgn, msg.data, msg.data.length))
      } else {
        this.messageCb({ pgn, length: msg.data.length, data: msg.data })

```
After which I modified the messageCb function to test the data output with COG:

```
let messageCb = (data) => {
    let jsonData = parser.parse(data, (err) => { if (err) console.error(err) })
    if (jsonData) {
       if (jsonData.canId == (167248384)) {console.log(jsonData.fields.COG)}

    }
}

```
So that part is now working well, although it is not appearing on my Garmin chartplotter as a NMEA2000 device. I fired a bunch of console.log() lines into the candevice.js file and it appears to be sending the isorequest and address claim and device information, but it's just not popping up on the garmin. It's also not sending anything over the CANbus when the chartplotter is plugged into the network, I'm pretty sure.

I'm also pretty keen to have a go at the custom PGN's!

**Comment by sbender9** [comment]
Definitely a bug! PRs welcome!

I would not use `canId` for that test, `PGN` would better since that device could send multiple PGNs.

And example of using custom PGNs: https://github.com/canboat/canboatjs/blob/master/test/customPgns.js

**Reply by sbender9** [reply]
An sk server plugin that does custom PGNs: https://github.com/sbender9/signalk-custom-pgn-example

**Comment by Aaarrrggghhhh** [comment]
Thanks for your help Scott

Did I do the PR correctly? I have never done one before.

I think I got most of the stuff going, the only thing I can't really check is the handshaking, which doesn't appear to be working but I don't have the equipment or knowledge to analyze exactly what is happening in that regard. 

For the benefit of whoever comes after me, this should be a working example which shows checking parsing incoming canbus data, utilising custom pgns and sending an example pgn on the canbus.


```
const canboatjs = require('@canboat/canboatjs')

const PropertyValues =  require('@signalk/server-api').PropertyValues

// I modified the example so the definition is in a different file for 
// convenience. If I knew what I was doing I would put it in a 
// correct filetype. It's literally copy pasted the PGNs array with 
// module.exports = { at the top
const definitions = require('./def.js')

const propertyValues = new PropertyValues();

  propertyValues.emitPropertyValue({
    timestamp: Date.now(),
    setter: 'customPgns',
    name: 'canboat-custom-pgns',
    value: definitions
  })

// Not sure about these lines of code. But it works?
var parser = new canboatjs.FromPgn({
    onPropertyValues: (name, cb) => {
      propertyValues.onPropertyValues(name, cb)
    }
  })

// This part takes the incoming raw canbus data and turns it into canboat objects
// I may add to this to make an array of objects recieved, and replace the objects
// in the array as they are recieved.
let messageCb = (data) => {
     
    let jsonData = parser.parse(data, (err) => { if (err) console.log(err) })
    if (jsonData) {       
        if (jsonData.pgn == 127250) { //Obviously change number to PGN required. 
            //Do whatever like console.log(jsonData.fields.Heading) in this example.
        }
    }
}

// I'm not sure this is working correctly, still not appearing on the Garmin
// chartplotter
let simpleCan = new canboatjs.SimpleCan({
    canDevice: 'can0',    //Obviously change this to the correct canbus address
    preferredAddress: 35,
    disableDefaultTransmitPGNs: true,
    transmitPGNs: [],
    addressClaim: {
        "Unique Number": 139725,
        "Manufacturer Code": 'Fusion Electronics',
        "Device Function": 130,
        "Device Class": 'Entertainment',
        "Device Instance Lower": 0,
        "Device Instance Upper": 0,
        "System Instance": 0,
        "Industry Group": 'Marine'
    },
    productInfo: {
        "NMEA 2000 Version": 1300,
        "Product Code": 667,
        "Model ID": "MS-UD650",
        "Software Version Code": "1.0",
        "Model Version": "1.0",
        "Model Serial Code": "123456",
        "Certification Level": 0,
        "Load Equivalency": 1
		    }
}, messageCb)

simpleCan.start()

// This is just a test message to check the sending over the canbus
testMessage=   {
    canId: 166793728,
    prio: 2,
    src: 35,
    dst: 255,
    pgn: 127250,
    timestamp: '2021-11-03T01:02:06.621Z',
    fields: {
      SID: 152,
      Heading: 4.674,
      Deviation: 0,
      Variation: 0.3624,
      Reference: 'True'
    },
    description: 'Vessel Heading'
  }

// send the test message every second so you can check it on connected
// device
setInterval(() => {
    simpleCan.sendPGN(testMessage);
}, 1000)

```

**Comment by phatpaul** [comment]
Thanks for the example, just what I was looking for.  But I don't see the sent testMessage on the bus.  Did you get this working?

I installed this from git clone and then `npm install`.  
Then I run `node examples/example2.js`

Ubuntu 24.04, Node v20.18.0, PEAK Pcan-usb (socketcan)
started with `sudo ip link set up can0 type can bitrate 250000 restart-ms 100`
